### PR TITLE
Common data store for the rules

### LINKED ISSUE #426: Fix cypress create rules flaky tests
Cypress tests for creating rules fail randomly. Fix the flakiness in the test. 

**Do you have any additional context?**
Looks like creating the rule from the test fail at some point and assertation fails after that trying to find that rule.

[cypress-videos-windows-latest (1).zip](https://github.com/opensearch-project/security-analytics-dashboards-plugin/files/10795326/cypress-videos-windows-latest.1.zip)
[cypress-screenshots-windows-latest (1).zip](https://github.com/opensearch-project/security-analytics-dashboards-plugin/files/10795327/cypress-screenshots-windows-latest.1.zip)

### LINKED ISSUE #473: [FEATURE] Common data store for the rules
**Is your feature request related to a problem?**
There is no common component to fetch and cache data for the rules.

**What solution would you like?**
Create a common data store for the rules, the store should be able to make requests to create/update/delete rules and also to cache the data after any request.

### PR DESCRIPTION
**Pull request by jovancacvetkovic**
### Description
Added common data store for the rules requests

### Issues Resolved
Closes #473 
Closes #426 

### Check List
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/pages/Rules/store/RulesStore.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+   * @param {{ [key: string]: string[] }} terms
+   * @param {any} query
+   * @param {boolean} invalidateCache
+   * @returns {Promise<RuleItemInfoBase[]>}
+   */
+  async getRules(
+    prePackaged: boolean,
+    terms?: { [key: string]: string[] },
+    query?: any,
+    invalidateCache: boolean = false
```
Comment (amsiglan) [inline public/pages/Rules/store/RulesStore.ts:96]: invalidation should happen only on select actions like create/update rule. Get should always return the cached value and does not need to provide an option for invalidation

File: public/pages/Rules/store/RulesStore.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+   * Makes the request to get pre-packaged or custom rules
+   * Pass `invalidateCache` to invalidate cache
+   *
+   * @param {boolean} prePackaged
+   * @param {{ [key: string]: string[] }} terms
+   * @param {any} query
+   * @param {boolean} invalidateCache
+   * @returns {Promise<RuleItemInfoBase[]>}
+   */
+  async getRules(
```
Comment (amsiglan) [inline public/pages/Rules/store/RulesStore.ts:92]: mark methods as private/public as appropriate

File: public/pages/Rules/components/RuleEditor/RuleEditorContainer.tsx [hunk 3]
```diff
@@ -14,7 +14,7 @@ import { RuleEditorFormModel, ruleEditorStateDefaultValue } from './RuleEditorFo
 import { mapFormToRule, mapRuleToForm } from './mappers';
 import { RuleEditorForm } from './RuleEditorForm';
 import { validateRule } from '../../utils/helpers';
-import { errorNotificationToast } from '../../../../utils/helpers';
+import { DataStore } from '../../../../store/DataStore';
 
 export interface RuleEditorProps {
```
Comment (amsiglan) [inline public/pages/Rules/components/RuleEditor/RuleEditorContainer.tsx:19]: Should we remove the ruleService prop now?

Comment (amsiglan) [reply public/pages/Rules/components/RuleEditor/RuleEditorContainer.tsx:19]: Similarly check in other places as well.

File: types/Rule.ts [hunk 4]
```diff
@@ -91,3 +94,33 @@ export interface DeleteRuleParams {
 }
 
 export interface DeleteRuleResponse {}
+
+export interface IRulesCache {
+  [key: string]: RuleItemInfoBase[] | [];
```
Comment (amsiglan) [inline types/Rule.ts:99]: Don't need to specify `| []` separately, an array can be empty for the specified type

File: types/Rule.ts [hunk 5]
```diff
@@ -91,3 +94,33 @@ export interface DeleteRuleParams {
...
+}
+
+export interface IRulesStore {
+  cache: IRulesCache;
```
Comment (amsiglan) [inline types/Rule.ts:103]: This doesn't need to be public so we can leave it out, no? Similarly for other properties which shouldn't be exposed like the invalidateCache method?

File: public/pages/Rules/store/RulesStore.ts (hunk 1 above)
Comment (jovancacvetkovic) [reply public/pages/Rules/store/RulesStore.ts]: Yup, set only for create/update/delete.

File: public/pages/Rules/components/RuleEditor/RuleEditorContainer.tsx (hunk 3 above)
Comment (jovancacvetkovic) [reply public/pages/Rules/components/RuleEditor/RuleEditorContainer.tsx:18]: All unnecessary rule service imports are removed.

File: types/Rule.ts (hunk 5 above)
Comment (jovancacvetkovic) [reply types/Rule.ts]: public/private done

File: public/pages/Alerts/components/AlertFlyout/AlertFlyout.tsx [hunk 6]
```diff
@@ -113,12 +108,23 @@ export class AlertFlyout extends React.Component<AlertFlyoutProps, AlertFlyoutSt
...
+          query: {
+            nested: {
+              path: 'rule',
+              query: {
+                terms: {
+                  _id: ruleIds,
+                },
+              },
+            },
+          },
```
Comment (amsiglan) [inline public/pages/Alerts/components/AlertFlyout/AlertFlyout.tsx:123]: Shall we only pass the terms object `{ _id: ruleIds }` and have the rest as placeholder in the RulesStore?

Comment (amsiglan) [reply public/pages/Alerts/components/AlertFlyout/AlertFlyout.tsx:123]: I see we use same options with different terms in the CreateDetector.tsx as well

Comment (jovancacvetkovic) [reply public/pages/Alerts/components/AlertFlyout/AlertFlyout.tsx:123]: Let me check.

File: public/pages/Rules/containers/Rules/Rules.tsx [hunk 7]
```diff
@@ -120,7 +116,7 @@ export const Rules: React.FC<RulesProps> = (props) => {
         </EuiFlexItem>
         <EuiFlexItem>
           <EuiPanel>
-            <RulesTable loading={loading} ruleItems={ruleItems} showRuleDetails={setFlyoutData} />
+            <RulesTable loading={loading} ruleItems={allRules} showRuleDetails={setFlyoutData} />
```
Comment (amsiglan) [inline public/pages/Rules/containers/Rules/Rules.tsx:119]: Is `allRules` using same interface as the `ruleItems`?

**Review by amsiglan — APPROVED** [review]

File: public/pages/Rules/containers/Rules/Rules.tsx (hunk 7 above)
Comment (jovancacvetkovic) [reply public/pages/Rules/containers/Rules/Rules.tsx:119]: Yes, just a naming consistency, as here we get all the rules, pre-packaged and custom once, hence changed the naming to **allRules**

File: public/pages/Alerts/components/AlertFlyout/AlertFlyout.tsx (hunk 6 above)
Comment (jovancacvetkovic) [reply public/pages/Alerts/components/AlertFlyout/AlertFlyout.tsx]: code base updated.

File: public/pages/Rules/store/RulesStore.ts [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+  /**
+   * Returns all rules, custom and pre-packaged
+   *
+   * @method getAllRules
+   * @param {terms?: { [key: string]: string[] }} [terms]
+   * @returns {Promise<RuleItemInfoBase[]>}
+   */
+  public async getAllRules(terms?: { [key: string]: string[] }): Promise<RuleItemInfoBase[]> {
+    let prePackagedRules = await this.getRules(false, terms);
```
Comment (amsiglan) [inline public/pages/Rules/store/RulesStore.ts:68]: why are we passing `false` for `prePackaged` and `true` for `custom`?

Comment (jovancacvetkovic) [reply public/pages/Rules/store/RulesStore.ts:68]: prePackaged is pat of the request query, I could add two more methods, getPackagedRules, and getCustomRules which would have these params already set, just to be more convenient.

**Review by amsiglan — APPROVED** [review]

**Review by AWSHurneyt — APPROVED** [review]

**Pull request merged**